src/sympc/protocol/falcon/falcon.py

```diff
@@ -73,87 +77,295 @@ def __eq__(self, other: Any) -> bool:
 
-        Performs Falcon's mul implementation, gets and reshares mul results and distributes shares.
-        This needs to be improved in future, it relies on orchestrator being a trusted third party.
-        Falcon, requires parties to be able to communication between each other.
-
         Args:
             x (MPCTensor): Secret
             y (MPCTensor): Another secret
             session (Session): Session the tensors belong to
+            op_str (str): Operation string.
+            kwargs_ (Dict[Any, Any]): Kwargs for some operations like conv2d
 
         Returns:
-            shares (ReplicatedSharedTensor): Shares in terms of ReplicatedSharedTensor.
+            result(MPCTensor): Result of the operation.
 
         Raises:
             ValueError: Raised when number of parties are not three.
-            NotImplementedError: Raised when implementation not present
-
+            ValueError : Raised when invalid security_type is provided.
         """
         if len(session.parties) != 3:
             raise ValueError("Falcon requires 3 parties")
 
         result = None
 
         if session.protocol.security_type == "semi-honest":
-            args = list(zip(x.share_ptrs, y.share_ptrs))
+            result = Falcon.mul_semi_honest(x, y, session, op_str, **kwargs_)
+        elif session.protocol.security_type == "malicious":
+            result = Falcon.mul_malicious(x, y, session, op_str, **kwargs_)
+        else:
+            raise ValueError("Invalid security_type for Falcon multiplication")
 
-            z_shares_ptrs = parallel_execution(
-                Falcon.compute_zvalue_and_add_mask, session.parties
-            )(args)
+        result = ABY3.truncate(result, session)
 
-            result = MPCTensor(shares=z_shares_ptrs, session=x.session)
-            result.shape = MPCTensor._get_shape("mul", x.shape, y.shape)  # for prrs
-            result = ABY3.truncate(result, session)
+        return result
 
-        else:
-            raise NotImplementedError(
-                f"mult operation not implemented for {session.protocol.security_type} setting"
-            )
+    @staticmethod
+    def mul_semi_honest(
+        x: MPCTensor,
+        y: MPCTensor,
+        session: Session,
+        op_str: str,
+        truncate: bool = True,
+        **kwargs_: Dict[Any, Any],
+    ) -> MPCTensor:
+        """Falcon semihonest multiplication.
+
+        Performs Falcon's mul implementation, add masks and performs resharing.
+
+        Args:
+            x (MPCTensor): Secret
+            y (MPCTensor): Another secret
+            session (Session): Session the tensors belong to
+            op_str (str): Operation string.
+            truncate (bool) : Applies truncation on the result if set.
+            kwargs_ (Dict[Any, Any]): Kwargs for some operations like conv2d
+
+        Returns:
+            MPCTensor: Result of the operation.
+        """
+        result = None
+
+        args = [
```

Comment:
MAYBE: `args = list(zip(x.share_ptrs, y.share_ptrs))` (I am not sure if it is more compact and clear?)

Reply:
I included the op_str like you suggested,
<code>
args = [
            [x_share, y_share, op_str]
            for x_share, y_share in zip(x.share_ptrs, y.share_ptrs)
        ]
</code>